### PR TITLE
[Merged by Bors] - feat(SetTheory/Ordinal/Nimber): Initial development of nimbers

### PR DESCRIPTION
**Pull request by vihdzp**
For the moment, we only prove that they form an additive commutative group.

---

I've proven that they form a field in my branch [vi.nimber_mul](https://github.com/leanprover-community/mathlib4/tree/vi.nimber_mul).

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (Command-Master) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:208]: This could also be proved by copying `small_Iio`, `small_Iic` and `not_small_ordinal` to nimbers and then stating that this set is too small,
```lean
theorem add_nonempty (a b : Nimber.{u}) :
    {x | (∃ a' < a, a' + b = x) ∨ ∃ b' < b, a + b' = x}ᶜ.Nonempty := by
  simp_rw [Set.nonempty_compl, Set.setOf_or, ← Set.mem_Iio, ← Set.image.eq_1]
  apply_fun (fun a : Set Nimber ↦ Small.{u} a)
  have : Small.{u} ↑((· + b) '' Set.Iio a ∪ (a + ·) '' Set.Iio b) := inferInstance
  simp [this, small_congr (Equiv.Set.univ _)]
  exact not_small_nimber
```
I'm not sure what's cleaner.

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (Command-Master) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:261]:
```suggestion
protected theorem add_comm (a b : Nimber) : a + b = b + a := by
  rw [add_def, add_def]
  simp_rw [or_comm]
  congr! 7 <;>
  (rw [and_congr_right_iff]; intro; rw [Nimber.add_comm])
termination_by (a, b)
```

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (Command-Master) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:347]: Perhaps it's more symmetrical to take `a + b + c ≠ 0` instead of `a ≠ b + c`?

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (Command-Master) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:345]: Shouldn't `add_cancel_right` and `add_cancel_left` be `@[simp]`?

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (Command-Master) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:263]: I don't think this should be a simp lemma. For comparison, `sub_eq_zero` isn't `@[simp]`.

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (Command-Master) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:264]: Perhaps this should be called `add_eq_zero`, to match `sub_eq_zero`?

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (vihdzp) [reply Mathlib/SetTheory/Ordinal/Nimber.lean:347]: I agree, but this lemma is meant to match [`Nat.xor_trichotomy`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Nat/Bitwise.html#Nat.xor_trichotomy) which uses this same assumption.

What do you think about refactoring that theorem too?

Comment (Command-Master) [reply Mathlib/SetTheory/Ordinal/Nimber.lean:341]: That sounds like a good idea

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (vihdzp) [reply Mathlib/SetTheory/Ordinal/Nimber.lean:208]: `blsub₂` is some code stink that should eventually be removed, all it really does is codify the fact that `Iio o × Iio o` is a small type. So I'm glad to replace it by an argument using `Small`.

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (vihdzp) [reply Mathlib/SetTheory/Ordinal/Nimber.lean:339]: Sure thing.

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (Command-Master) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:143]:
```suggestion
theorem not_small_nimber : ¬ Small.{u} Nimber.{max u v} :=
```
This is `Nimber.not_small_nimber` compared to `_root_.not_small_ordinal`, I think this should be outside the namespace.

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (Command-Master) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:141]: Could you also transfer `Ordinal.small_Iic`?

**Review by Command-Master — APPROVED** [review]

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (vihdzp) [reply Mathlib/SetTheory/Ordinal/Nimber.lean]: In fact, I could make the statement even more symmetrical by changing it to
```
b + c < a ∨ c + a < b ∨ a + b < c
```

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (vihdzp) [reply Mathlib/SetTheory/Ordinal/Nimber.lean:141]: Sure. It'd be nice to be able to infer `Small (Set.Iic a)` (and the other intervals) from `Small (Set.Iio a)` and `NoMaxOrder`.

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (b-mehta) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:339]: Let's move this TODO to the list of others at the top

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (b-mehta) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:36]: It would also be nice to relate the bitwise `xor` definition on naturals to the addition you've defined here. Something like, doing `xor` on naturals then embedding into nimbers is the same as embedding into nimbers then adding

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (b-mehta) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:222]:
```suggestion
  simpa [this, small_congr (Equiv.Set.univ _)] using not_small_nimber
```
looks like it might work, and eliminates a non-terminal simp.
If it doesn't work, then try to eliminate the non-terminal simp in some other way

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (b-mehta) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:312]: This seems worth investigating further - can leave that to another pr though

**Review by b-mehta — APPROVED** [review]
maintainer merge

File: Mathlib/SetTheory/Ordinal/Nimber.lean
Comment (vihdzp) [reply Mathlib/SetTheory/Ordinal/Nimber.lean:36]: This shouldn't be too difficult to prove using induction plus [`Nat.lt_xor_cases`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Nat/Bitwise.html#Nat.lt_xor_cases), but it does require porting some of the API of `Ordinal.natCast` over, so I'll do it in a separate PR.

Something that might also be good to do in said PR is introduce the notation `∗x` for `Ordinal.toNimber x`.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
